File: 프로그래머스/2/43165. 타겟 넘버/타겟 넘버.cpp
#include <string>
#include <vector>
#include <deque>

using namespace std;

int solution(vector<int> numbers, int target) {
    int answer = 0;
    deque<int> queue;
    int index = 1;
    queue.push_front(target);
    while(numbers.size()>0){
        int number = numbers.back();
        numbers.pop_back();
        bool isEnd = (numbers.size()==0);
        for(int i=0;i<index;i++){
            int node = queue[0];
            queue.pop_front();
            if(isEnd)
            {
                if(node-number==0)answer++;
                if(node+number==0)answer++;
            }
            else{
            queue.push_back(node-number);
            queue.push_back(node+number);
            }
        }
        index*=2;        
    }
    
    return answer;
}

File: 프로그래머스/2/43165. 타겟 넘버/타겟 넘버.py
from collections import deque
def solution(numbers, target):
    answer = 0
    search = deque()
    search.append([0,0,1])
    search.append([0,0,-1])
    while len(search)>0:
        current = search.popleft()
        value = current[0]+numbers[current[1]]*current[2]
        if current[1]==len(numbers)-1:
            if value==target:
                answer+=1
        else:
            search.append([value,current[1]+1,1])
            search.append([value,current[1]+1,-1])
    return answer